k2/csrc/fsa_test.cc

```diff
@@ -0,0 +1,316 @@
+  //
+  // 20-byte header             (20)
+  // 44-byte padding            (64)
+  // 12-byte state_offset_array (76)
+  // 52-byte padding            (128)
+  // 120-byte arc_indexes_array (248)
+  // 4-byte padding             (252) -> to be multiple of sizeof(Arc)
+  // 108-byte arcs_array        (360)
+  EXPECT_EQ(bytes, 360u);
+}
+
+TEST(CfsaVec, Empty) {
+  Cfsa cfsa;
+  std::vector<Cfsa> cfsas;
+  size_t bytes = GetCfsaVecSize(cfsas);
+  std::unique_ptr<void, decltype(&MemFree)> data(MemAlign(bytes, 64), &MemFree);
+
+  CreateCfsaVec(cfsas, data.get(), bytes);
+
+  CfsaVec cfsa_vec(bytes / sizeof(int32_t), data.get());
+  EXPECT_EQ(cfsa_vec.NumFsas(), 0);
+}
+
+TEST(CfsaVec, OneEmptyCfsa) {
+  Cfsa cfsa;
+  std::vector<Cfsa> cfsas = {cfsa};
+  size_t bytes = GetCfsaVecSize(cfsas);
+  std::unique_ptr<void, decltype(&MemFree)> data(MemAlign(bytes, 64), &MemFree);
+
+  CreateCfsaVec(cfsas, data.get(), bytes);
+
+  CfsaVec cfsa_vec(bytes / sizeof(int32_t), data.get());
+  EXPECT_EQ(cfsa_vec.NumFsas(), 1);
+}
+
+TEST(CfsaVec, OneNoneEmptyCfsa) {
+  std::string s = R"(
+      0 1 10
+      0 2 2
+      1 3 3
+      2 3 3
+      3 4 -1
+      2 4 -1
+      4
+  )";
+
+  auto fsa = StringToFsa(s);
+
+  Cfsa cfsa(*fsa);
+  std::vector<Cfsa> cfsas = {cfsa};
+  size_t bytes = GetCfsaVecSize(cfsas);
+  std::unique_ptr<void, decltype(&MemFree)> data(MemAlign(bytes, 64), &MemFree);
+
+  CreateCfsaVec(cfsas, data.get(), bytes);
+
+  CfsaVec cfsa_vec(bytes / sizeof(int32_t), data.get());
+  EXPECT_EQ(cfsa_vec.NumFsas(), 1);
+
+  Cfsa f = cfsa_vec[0];
+  EXPECT_EQ(f, cfsa);
+}
+
+TEST(CfsaVec, TwoNoneEmptyCfsa) {
+  std::string s1 = R"(
+      0 1 10
+      0 2 2
+      1 3 3
+      2 3 3
+      3 4 -1
+      2 4 -1
+      4
+  )";
+
+  auto fsa1 = StringToFsa(s1);
+
+  Cfsa cfsa1(*fsa1);
+
+  std::string s2 = R"(
+      0 1 10
+      0 2 2
+      1 3 3
+      2 3 3
+      3 10 -1
+      2 4 3
+      4 10 -1
+      10
+  )";
+
+  auto fsa2 = StringToFsa(s2);
+
+  Cfsa cfsa2(*fsa2);
+
+  {
+    // both fsa are not empty
+    std::vector<Cfsa> cfsas = {cfsa1, cfsa2};
+    size_t bytes = GetCfsaVecSize(cfsas);
+    std::unique_ptr<void, decltype(&MemFree)> data(MemAlign(bytes, 64),
+                                                   &MemFree);
+
+    CreateCfsaVec(cfsas, data.get(), bytes);
+
+    CfsaVec cfsa_vec(bytes / sizeof(int32_t), data.get());
+    EXPECT_EQ(cfsa_vec.NumFsas(), 2);
+
+    Cfsa f = cfsa_vec[0];
+    EXPECT_EQ(f, cfsa1);
+
+    Cfsa g = cfsa_vec[1];
+    EXPECT_EQ(g, cfsa2);
+  }
+
+  {
+    // the first fsa is empty
+    Cfsa cfsa;
+    std::vector<Cfsa> cfsas = {cfsa, cfsa2};
+    size_t bytes = GetCfsaVecSize(cfsas);
+    std::unique_ptr<void, decltype(&MemFree)> data(MemAlign(bytes, 64),
+                                                   &MemFree);
+
+    CreateCfsaVec(cfsas, data.get(), bytes);
+
+    CfsaVec cfsa_vec(bytes / sizeof(int32_t), data.get());
+    EXPECT_EQ(cfsa_vec.NumFsas(), 2);
+
+    Cfsa f = cfsa_vec[0];
```

Comment:
using const reference instead of copy assignment? (even if the copy is cheap for Cfsa)

Reply:
cfsa_vec operator [] doesn't return a reference (last I checked, anyway).  Hopefully compiler will inline.

Reply:
mark as resolved.